# Building a Spam Filter with Naive Bayes

In this guided project, we are going to study the practical side of the Naive Bayes algorithm by building a spam filter for SMS messages.

To classify messages as spam or non-spam the computer:

- Learns how humans classify messages.
- Uses that human knowledge to estimate probabilities for new messages — probabilities for spam and non-spam.
- Classifies a new message based on these probability values — if the probability for spam is greater, then it classifies the message as spam. Otherwise, it classifies it as non-spam (if the two probability values are equal, then we may need a human to classify the message).

So our first task is to "teach" the computer how to classify messages. To do that, we will use the multinomial Naive Bayes algorithm along with a dataset of 5,572 SMS messages that are already classified by humans.

The dataset was put together by Tiago A. Almeida and José María Gómez Hidalgo, and it can be downloaded from [The UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection). The data collection process is described in more details on [this page](http://www.dt.fee.unicamp.br/~tiago/smsspamcollection/#composition), where you can also find some of the authors' papers.

In [81]:
import pandas as pd
import re

In [82]:
sms = pd.read_csv('SMSSpamCollection', sep='\t', header=None, names=['Label', 'SMS'])

In [83]:
#to see the number of rows and columns
sms.shape

(5572, 2)

In [84]:
sms.head()

Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [85]:
#'Ham' means non-spam. Below we can see what percentage of the SMS are spam and non-spam

sms['Label'].value_counts(normalize=True)*100

ham     86.593683
spam    13.406317
Name: Label, dtype: float64

Above we read in the dataset and saw that about 87% of the messages are ham ("ham" means non-spam), and the remaining 13% are spam. Now that we have become a bit familiar with the dataset, we can move on to building the spam filter.

However, before creating it, it's very helpful to first think of a way of testing how well it works. When creating software, a good rule of thumb is that designing the test comes before creating the software. If we write the software first, then it's tempting to come up with a biased test just to make sure the software passes it.

Once our spam filter is done, we will need to test how good it is with classifying new messages. To test the spam filter, we are first going to split our dataset into two categories:

- A training set, which we will use to "train" the computer how to classify messages.
- A test set, which we will use to test how good the spam filter is with classifying new messages.

Wea re going to keep 80% of our dataset for training, and 20% for testing (we want to train the algorithm on as much data as possible, but we also want to have enough test data). The dataset has 5,572 messages, which means that:

- The training set will have 4,458 messages (about 80% of the dataset).
- The test set will have 1,114 messages (about 20% of the dataset).

To better understand the purpose of putting a test set aside, let's begin by observing that all 1,114 messages in our test set are already classified by a human. When the spam filter is ready, we're going to treat these messages as new and have the filter classify them. Once we have the results, we'll be able to compare the algorithm classification with that done by a human, and this way we'll see how good the spam filter really is.

For this project, our goal is to create a spam filter that classifies new messages with an accuracy greater than 80% — so we expect that more than 80% of the new messages will be classified correctly as spam or ham (non-spam).

We'll come back to testing toward the end of this guided project, but for now, let's create a training and a test set. We're going to start by randomizing the entire dataset to ensure that spam and ham messages are spread properly throughout the dataset.

In [86]:
#We randomize the data set
random = sms.sample(frac=1, random_state=1)
random.shape

(5572, 2)

In [87]:
#We split the data set into two: 
#80% into a training set we will use to teach the computer
#20% into a test set we will use to test our algorithm.

training = random[:4458].reset_index(drop=True)
test = random[4458:].reset_index(drop=True)

In [88]:
training['Label'].value_counts(normalize=True)*100

ham     86.54105
spam    13.45895
Name: Label, dtype: float64

In [89]:
test['Label'].value_counts(normalize=True)*100

ham     86.804309
spam    13.195691
Name: Label, dtype: float64

## Data clean up

In [90]:
training['SMS'] = training['SMS'].str.replace('\W', ' ')
training['SMS'] = training['SMS'].str.lower()
training.head()

Label                                                SMS
0   ham                       yep  by the pretty sculpture
1   ham      yes  princess  are you going to make me moan 
2   ham                         welp apparently he retired
3   ham                                            havent 
4   ham  i forgot 2 ask ü all smth   there s a card on ...

In [91]:
training['SMS'] = training['SMS'].str.split()

In [92]:
vocabulary = []

for row in training['SMS']:
    for word in row:
        if word not in vocabulary:
            vocabulary.append(word)
        else:
            pass
        
print(len(vocabulary))

7783


In [93]:
word_counts_per_sms = {unique_word: [0] *
len(training['SMS']) for unique_word in vocabulary}

for index, sms in enumerate(training['SMS']):
    for word in sms:
        word_counts_per_sms[word][index] += 1
        
    

In [58]:
word_count = pd.DataFrame(word_counts_per_sms)
word_count.head()

yep  by  the  pretty  sculpture  yes  princess  are  you  going  ...  \
0    1   1    1       1          1    0         0    0    0      0  ...   
1    0   0    0       0          0    1         1    1    1      1  ...   
2    0   0    0       0          0    0         0    0    0      0  ...   
3    0   0    0       0          0    0         0    0    0      0  ...   
4    0   0    0       0          0    0         0    0    0      0  ...   

   beauty  hides  secrets  n8  jewelry  related  trade  arul  bx526  wherre  
0       0      0        0   0        0        0      0     0      0       0  
1       0      0        0   0        0        0      0     0      0       0  
2       0      0        0   0        0        0      0     0      0       0  
3       0      0        0   0        0        0      0     0      0       0  
4       0      0        0   0        0        0      0     0      0       0  

[5 rows x 7783 columns]

In [59]:
training_set = pd.concat([training, word_count], axis=1)
training_set.head()

Label                                                SMS  yep  by  the  \
0   ham                  [yep, by, the, pretty, sculpture]    1   1    1   
1   ham  [yes, princess, are, you, going, to, make, me,...    0   0    0   
2   ham                    [welp, apparently, he, retired]    0   0    0   
3   ham                                           [havent]    0   0    0   
4   ham  [i, forgot, 2, ask, ü, all, smth, there, s, a,...    0   0    0   

   pretty  sculpture  yes  princess  are  ...  beauty  hides  secrets  n8  \
0       1          1    0         0    0  ...       0      0        0   0   
1       0          0    1         1    1  ...       0      0        0   0   
2       0          0    0         0    0  ...       0      0        0   0   
3       0          0    0         0    0  ...       0      0        0   0   
4       0          0    0         0    0  ...       0      0        0   0   

   jewelry  related  trade  arul  bx526  wherre  
0        0        0      0     0      0       0  
1        0        0      0     0      0       0  
2        0        0      0     0      0       0  
3        0        0      0     0      0       0  
4        0        0      0     0      0       0  

[5 rows x 7785 columns]

In [60]:
training_set['Label'].value_counts()

ham     3858
spam     600
Name: Label, dtype: int64

In [61]:
p_spam = 600/(3858+600)
p_ham = 3858/(3858+600)

In [62]:
spam = training_set[training_set['Label'] == 'spam']
ham = training_set[training_set['Label'] == 'ham']

In [63]:
spam_number_words = spam['SMS'].apply(len)
n_spam = spam_number_words.sum()

In [64]:
ham_number_words = ham['SMS'].apply(len)
n_ham = ham_number_words.sum()

In [65]:
n_vocabulary = len(vocabulary)
alpha = 1

In [66]:
parameters_spam = {}
parameters_ham = {}

for word in vocabulary:
    parameters_spam[word] = 0
    parameters_ham[word] = 0

In [67]:
for word in vocabulary:
    n_word_given_spam = spam[word].sum()
    n_word_given_ham = ham[word].sum()
    p_word_given_spam = (n_word_given_spam+alpha)/(n_spam + alpha*n_vocabulary)
    p_word_given_ham = (n_word_given_ham+alpha)/(n_ham + alpha*n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    parameters_ham[word] = p_word_given_ham

In [68]:
def classify(message):
    message = re.sub('\W', ' ', message)
    message = message.lower()
    message = message.split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham
    
    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
            
    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)
    
    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal proabilities, have a human classify this!')

In [69]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.3481290211300841e-25
P(Ham|message): 1.9368049028589875e-27
Label: Spam


In [70]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.4372375665888117e-25
P(Ham|message): 3.687530435009238e-21
Label: Ham


In [71]:
def classify_test_set(message):

    message = re.sub('\W', ' ', message)
    message = message.lower()
    message = message.split()

    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]

        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]

    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

In [72]:
test['predicted'] = test['SMS'].apply(classify_test_set)

In [73]:
test.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

In [74]:
correct = 0
total = len(test)

In [75]:
for row in test.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct += 1

In [76]:
accuracy = correct/total

In [77]:
print('Correct:', correct)
print('Incorrect:', total-correct)
print('Accuracy:{:.2f}%'.format(accuracy*100))

Correct: 1100
Incorrect: 14
Accuracy:98.74%


## Conclusion

In this project, we managed to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. The filter had an accuracy of 98.74% on the test set, which is an excellent result. We initially aimed for an accuracy of over 80%, but we managed to do way better than that.

In [78]:
inacurate = test[test['Label'] != test['predicted']]

In [79]:
inacurate.head()

Label                                                SMS predicted
114  spam  Not heard from U4 a while. Call me now am here...       ham
135  spam  More people are dogging in your area now. Call...       ham
152   ham                  Unlimited texts. Limited minutes.      spam
159   ham                                       26th OF JULY      spam
284   ham                             Nokia phone is lovly..      spam